File: benchmarks/regex-redux/regex-redux.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
)

func main() {
	fileName := "25000_in"
	if len(os.Args) > 1 {
		fileName = os.Args[1]
	}

	f, err := os.Open(fileName)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	bytes, err := ioutil.ReadAll(f)
	if err != nil {
		panic(err)
	}
	originalLen := len(bytes)

	// Clean the input - remove headers and newlines
	cleanRE := regexp.MustCompile("(>[^\n]+)?\n")
	bytes = cleanRE.ReplaceAllLiteral(bytes, []byte(""))
	cleanedLen := len(bytes)

	// Variant patterns to count
	variants := []string{
		"agggtaaa|tttaccct",
		"[cgt]gggtaaa|tttaccc[acg]",
		"a[act]ggtaaa|tttacc[agt]t",
		"ag[act]gtaaa|tttac[agt]ct",
		"agg[act]taaa|ttta[agt]cct",
		"aggg[acg]aaa|ttt[cgt]ccct",
		"agggt[cgt]aa|tt[acg]accct",
		"agggta[cgt]a|t[acg]taccct",
		"agggtaa[cgt]|[acg]ttaccct",
	}

	// Count each variant
	for _, pattern := range variants {
		re := regexp.MustCompile(pattern)
		matches := re.FindAll(bytes, -1)
		count := len(matches)
		fmt.Printf("%s %d\n", pattern, count)
	}

	// Substitutions
	substitutions := []struct {
		pattern     string
		replacement string
	}{
		{"tHa[Nt]", "<4>"},
		{"aND|caN|Ha[DS]|WaS", "<3>"},
		{"a[NSt]|BY", "<2>"},
		{"<[^>]*>", "|"},
		{`\|[^|][^|]*\|`, "-"},
	}

	for _, sub := range substitutions {
		re := regexp.MustCompile(sub.pattern)
		bytes = re.ReplaceAll(bytes, []byte(sub.replacement))
	}

	fmt.Printf("\n%d\n%d\n%d\n", originalLen, cleanedLen, len(bytes))
}

File: benchmarks/regex-redux/regex-redux.rb
MATCHERS = [
  /agggtaaa|tttaccct/,
  /[cgt]gggtaaa|tttaccc[acg]/,
  /a[act]ggtaaa|tttacc[agt]t/,
  /ag[act]gtaaa|tttac[agt]ct/,
  /agg[act]taaa|ttta[agt]cct/,
  /aggg[acg]aaa|ttt[cgt]ccct/,
  /agggt[cgt]aa|tt[acg]accct/,
  /agggta[cgt]a|t[acg]taccct/,
  /agggtaa[cgt]|[acg]ttaccct/
]

SUBSTITUTIONS = [
  [/tHa[Nt]/, '<4>'],
  [/aND|caN|Ha[DS]|WaS/, '<3>'],
  [/a[NSt]|BY/, '<2>'],
  [/<[^>]*>/, '|'],
  [/\|[^|][^|]*\|/, '-']
]

file_name = ARGV[0] || "25000_in"
seq = File.read(file_name)
original_len = seq.size

# Clean the input
seq.gsub!(/>.*\n|\n/, '')
cleaned_len = seq.size

# Count variants
MATCHERS.each do |matcher|
  count = 0
  seq.scan(matcher) { count += 1 }
  puts "#{matcher.source} #{count}"
end

# Apply substitutions
SUBSTITUTIONS.each do |pattern, replacement|
  seq.gsub!(pattern, replacement)
end

puts
puts original_len
puts cleaned_len
puts seq.size
